src/liballoc/string.rs

```diff
@@ -647,6 +648,30 @@ impl String {
         decode_utf16(v.iter().cloned()).map(|r| r.unwrap_or(REPLACEMENT_CHARACTER)).collect()
     }
 
+    /// Decomposes a `String` into its raw components.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(vec_into_raw_parts)]
+    /// let s = String::from("hello");
+    ///
+    /// let (ptr, len, cap) = s.into_raw_parts();
+    ///
+    /// let rebuilt = unsafe { String::from_raw_parts(ptr, len, cap) };
+    /// assert_eq!(rebuilt, "hello");
+    /// ```
+    #[unstable(feature = "vec_into_raw_parts", reason = "new API", issue="0")]
+    pub fn into_raw_parts(self) -> (*mut u8, usize, usize) {
+        let ptr = self.as_ptr() as *mut u8;
```

Comment:
Quoting the `as_ptr` docs:

> The caller must ensure that the returned pointer is never written to.

I think `as_mut_ptr` is what you actually want.

Reply:
Hm, looks like `DerefMut` might be missing for `String`?
Still, the current implementation is invalid. `from_raw_parts` expects a ptr you can write to.

Reply:
For this `String` case you may as well just do `self.vec.into_raw_parts()`.

Reply:
You know, I was surprised by this code, but I copied it from the `from_raw_parts` right below! I can update that in a separate commit first, then pull the changes into this doc.

Reply:
Ouch... good catch! Yes that example there is wrong.

We should really [get Miri to run doctests](https://github.com/rust-lang/miri/issues/584)...